### PR TITLE
Added OpenFX Plugins

### LINKED ISSUE #1117: Write OpenFX plug-ins for OCIO v2
Particularly as we wait for applications to integrate OCIO v2, it would be awesome to have OpenFX plug-ins that use the new library to aid testing.  

Looking for a volunteer to take this on!   The Autodesk team will be happy to consult and it could be a nice learning opportunity for someone.

Also, please note that there are some existing OpenFX plug-ins for OCIO v1 that could be studied as examples, however we are looking for new plug-ins to be written from scratch that could be included in the OCIO repo.
Natron:
https://github.com/MrKepzie/openfx-io/tree/master/OCIO
TuttleOFX:
https://github.com/tuttleofx/TuttleOFX/tree/develop/plugins/image/process/color/OCIO

We would especially like to have plug-ins to wrap ColorSpaceTransform, DisplayViewTransform, and FileTransform.  The goal would be to leverage the new v2 features where possible.  These could also serve as a nice reference for application developers looking to upgrade to v2.

Resources for OpenFX
-----------------------
The Open Effects Association website:
http://openeffects.org/

OpenFX Programming Guide on ReadTheDocs:
https://openfx.readthedocs.io/en/doc/Guide/index.html

OpenFX repo on GitHub, which contains example plug-ins:
https://github.com/ofxa/openfx

List  of host apps that may be used to test the plug-ins:
https://en.wikipedia.org/wiki/OpenFX_(API)

Recommended approach:

1) Download the official OpenFX example plug-ins, compile them, and test that they work in your host app of choice.
2) Read the Programming Guide.
3) Start with one of the example programs such as openfx/Examples/Invert or Basic and start modifying it to call OCIO.

### PR DESCRIPTION
**Pull request by EmperorYP7**
Fixes #1117

**Description:** Here's my implementation for the OpenFX plugins for OCIO v2. The current state of my work is hardly a "working" prototype plugin for `ColorSpaceTransform`, `FileTransform` and `DisplayViewTransform` (Not yet completely implemented it). This PR will keep track of the progress of the plugins as I develop it, making necessary changes upfront as opposed to changing it at a later stage of development.

ColorSpace Transform Plugin in Natron :

Screenshot 2021-01-20 at 5 32 59 PM

### CODE REVIEW

**Comment by michdolan** [comment]
@EmperorYP7 please let us know if you have implementation questions at any point. This work plays a very important role in demonstrating OCIO v2 functionality to application implementers so is of high priority for us to support you on. Thanks for your work on this.

I'll try to find some cycles soon to pull down your source and do a build. I recommend targeting Resolve as an initial test case since it's free and stable. Please don't hesitate to reach out.

**Comment by EmperorYP7** [comment]
That sounds great to me sir!

Doug Walker sir has made me clear about the approach I should be taking for this. I'm all ears to the inputs given by the mentors and community. :smile:

**Comment by EmperorYP7** [comment]
@michdolan sir, I am facing some problems in building and debugging the plugins in DaVinci Resolve. My current approach is to replace one of the example plugins of the [official OFX repo](https://github.com/ofxa/openfx) with my plugin code and building it with the default build method provided in it. I don't think this is an optimal way of building. I'd be glad if you suggest me a better way to build it. Thanks!

File: vendor/openfx/PluginMain.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+#include "ofxImageEffect.h"
+#include "ofxMemory.h"
+#include "ofxMultiThread.h"
+
+#include <OpenColorIO/OpenColorIO.h>
+
+namespace OCIO = OCIO_NAMESPACE;
+
+#if defined __APPLE__ || defined linux || defined __FreeBSD__
```
[the remark this answers is not in the document]
Comment (EmperorYP7) [reply vendor/openfx/PluginMain.cpp:18]: Oops.. missed it! Thanks for pointing out. 😄

**Comment by michdolan** [comment]
Apologies @EmperorYP7 for the long delay in helping you with this. I'm going to pull down your branch and try to get one of these plugins functional in Resolve, and can then make a PR back to your branch so you can continue work on these.

**Comment by michdolan** [comment]
I reworked the CMake side of this and moved the OpenFX install from a submodule to OCIO's INSTALL_EXP_PACKAGES system, and have the OFX bundle building for Linux and Windows. The plugin loads in Resolve, but is not exposed in the application yet as far as I can tell. I'll do some investigation there next:
image

**Comment by remia** [comment]
The plugin on your branch is available for me on Resolve OFX list, but it don't have any effect and don't expose parameters. Console says ```Create instance failed for plugin [opencolorio.ociocolorspace] with context [OfxImageEffectContextFilter_<...uuid...>]!!!!```.

I then tried the OCIOFileTransform plugin from this PR but that crashed Resolve as soon as activated.

There is sample plugins available in Resolve dev documentation folder by the way, they use a different API, OO based.

**Comment by michdolan** [comment]
Thanks for trying these out @remia . In my branch I currently have one of the OFX examples plugged in just to get something working, which I will then swap with OCIO functionality. But yes, as you observed the example (which is intended to invert the input image) is not working in Resolve currently (outside of showing up in the menu). I'll do some more digging into why this weekend.

**Comment by remia** [comment]
In case that might help, upon a first look at the `FileTransformPlugin` from @EmperorYP7, I got it to produce results in Resolve (I used the CMake refactor to compile it) making the following changes: https://gist.github.com/remia/6a79727169cbfb4962ab615a2e9d4263 . At least for this plugin, there is issues with parameters definition and the way we process the incoming image buffer without assigning the result.

**Comment by michdolan** [comment]
Thanks @remia for your suggestions. I will be curious to get your thoughts on #1371 as well.

@EmperorYP7 I reworked the build quite a bit and in my quest to get a plugin functional in Resolve I discovered the OpenFX Support Library, a bundled C++ wrapper around the C API. This is what the OFX plugin example that ships with Resolve uses, and that led me to follow that and other examples in the OpenFX repo resulting in the work in #1371 after a number of iterations. By the time I got to that point, and some abstraction to support reuse of the lower-level pieces of the plugin design, our branches had diverged to the point where it made more sense to make a new PR to the OCIO repo instead of to your fork, as I had originally intended.

There is still work to be done, as I have only implemented the simplest OCIOColorSpace case, and the other plugins you developed here are not accounted for. There is also opportunity to explore implementation of the OCIO GPU renderer within the plugin. I'm going to suggest we close this work and refocus our efforts on the C++ wrapper approach. If you have time it would be great to adapt your other plugins into that framework, which should allow you to focus on the implementation more. Thanks again for getting this effort started!

**Comment by EmperorYP7** [comment]
@michdolan no problem. I am occupied with other open-source projects and university assignments for now. Will work on it whenever I get some time. Thanks!

**Pull request closed without merging**